File: typescript/controllers/GameController.ts
import express, { Request, Response } from 'express';
import path from 'path';
import session from 'express-session';
import { WordSelectionService } from '../services/WordSelectionService';
import { GameStatus } from '../models/GameStatus';
import { Word } from '../models/Word';

interface GuessHistory {
    guess: string;
    result: string;
}

class GameController {

    public path = '/game'
    public router = express.Router();
    public workshopMode = false;
    public basePath = "";

    constructor(workshopMode: boolean, basePath: string) { 
        this.workshopMode = workshopMode;
        this.basePath = basePath;
        this.initializeRoutes();
    }

    public initializeRoutes() {

        this.router.get(this.path, (req: Request, res: Response) => {
            const WORD = new Word(new WordSelectionService().getWord());
            const attempts = 0;
            const user = req.query.user as string;
            
            if (!user) {
                res.redirect('/');
                return;
            }

            // Initialize session data
            (req as any).session = (req as any).session || {};
            (req as any).session.guessHistory = [];
        
            res.render('game.ejs', {
                message: 'Make your first guess!',
                attempts,
                status: GameStatus.INPROGRESS,
                user,
                guessHistory: [],
                workshopMode: this.workshopMode,
                basePath: this.basePath
            });
        });

        this.router.post(this.path, (req: Request, res: Response) => {
            const WORD = new Word(new WordSelectionService().getWord());
            let attempts = this.getAttempts(req);
            const guess = req.body.guess;
            const user = req.body.user || 'Default';
            attempts++;
            const result = WORD.getInfo(guess);
            var message = '';
            var status = 'UNKNOWN';
            
            // Initialize session if not exists
            (req as any).session = (req as any).session || {};
            const guessHistory: GuessHistory[] = (req as any).session.guessHistory || [];
            
            // Add current guess to history
            guessHistory.push({ guess, result });
            (req as any).session.guessHistory = guessHistory;
            
            // log the value of WORD and guess
            console.log(`Word: ${WORD.getWord()}, Guess: ${guess}`);

            if (WORD.isCorrect(guess)) {
                message = 'Congratulations! You guessed correctly.';
                status = GameStatus.SUCCESS;
                
            } else if (attempts >= 5) {
                message = 'Game Over! You have used all your attempts.';
                status = GameStatus.FAILED;
            } else {
                message = 'Try again. Your next guess:';
                status = GameStatus.INPROGRESS;
            }
            res.render('game.ejs', { message, attempts, guess, result, status, 
                user, 
                guessHistory,
                workshopMode: this.workshopMode,
                basePath: this.basePath }
            );
        });
    }

    private getAttempts(req: Request): number {
        try {
            const attempts = req.body.attempts;
            if (attempts === undefined) {
                return 0;
            }
            return parseInt(attempts);
        } catch (e) {
            console.log(e);
            return 0;
        }
    }
}

export = GameController;

File: python/Qwords/controllers/GameController.py
from flask import Blueprint, render_template, request, session, url_for, redirect

from Qwords.services.WordSelectionService import WordSelectionService
from Qwords.models.GameStatus import GameStatus
from Qwords.models.Word import Word
from Qwords.errors import InvalidWordError, GameSessionError
from Qwords.utils.url_utils import get_prefixed_url

game_page = Blueprint('game_page', __name__, template_folder='../templates')

def get_attempts(request: request) -> int: # type: ignore
    try:
        attempts = int(request.form.get('attempts', 0))
        return attempts
    except (ValueError, KeyError):
        return 0

@game_page.route('/game', methods=['GET', 'POST'])
def load_game():
    """
    Handle game load and gameplay logic.
    
    GET:
        Returns the game page with current state.
    POST:
        Process a word guess and update game state.
        
    Returns:
        Rendered game template with updated state.
        
    Raises:
        GameSessionError: If max attempts reached or username not set
        InvalidWordError: If invalid word provided
    """
    
    # Print out all request args
    print(request.args)
    print (request.form)
    print(dir(game_page))

    user=request.args.get('user', 'Guest')
    game_url = get_prefixed_url('game')
    home_url = get_prefixed_url('')
    
    # Handle GET request
    if request.method == 'GET':
        return render_game_state(
            message="Make your first guess!",
            attempts=0,
            user=user,
            status=GameStatus.INPROGRESS
        )
    
    attempts = get_attempts(request)
    
    # Check max attempts
    if attempts >= 5:
        raise GameSessionError("Sorry, you've reached the maximum number of attempts.")

    # Initialize word if not exists
    if not hasattr(load_game, 'current_word') or load_game.current_word is None:
        #debug log word
        print("Initializing word")
        load_game.current_word = WordSelectionService().get_random_word()
        print(load_game.current_word.value)

    # Handle POST request
    guess = request.form.get('guess', '').strip().lower()
    if not guess:
        raise InvalidWordError("Please enter a guess!")

    # Process guess
    attempts += 1
    print("Current word: " + load_game.current_word.value)
    if guess == load_game.current_word.value.lower():
        load_game.current_word = None  # Reset for next game
        return render_game_state(
            message="Congratulations! You've guessed the word!",
            attempts=attempts,
            user=user,
            status=GameStatus.SUCCESS
        )

    # Handle incorrect guess
    info = load_game.current_word.get_info(guess)
    if attempts >= 5:
        return render_game_state(
            message="Sorry, you've reached the maximum number of attempts.",
            attempts=attempts,
            user=user,
            status=GameStatus.FAILED
        )

    return render_game_state(
        message=f"Wrong guess! Try again.",
        attempts=attempts,
        guess=guess,
        result=[*info],
        user=user,
        status=GameStatus.INPROGRESS
    )

def render_game_state(message, attempts, status, guess=None, result=None, user="Guest"):
    """Helper function to render game template with consistent parameters."""
    return render_template(
        'game.html',
        message=message,
        attempts=attempts,
        guess=guess,
        result=result,
        status=status,
        home_url=get_prefixed_url(''),
        game_url=get_prefixed_url('game'),
        user=user
    )

    
@game_page.route('/new_game', methods=['POST'])
def new_game():
    """
    Start a new game session.
    
    POST Parameters:
        user (str): Username for the game session
        
    Returns:
        Redirect to the game page
    """
    user = request.form.get('user')
    if not user:
        raise GameSessionError("Username is required")
    return redirect(url_for('game_page.load_game', user=user))

